src/transport_udp.cpp

```diff
@@ -242,361 +520,480 @@ UDPTransport::fd_writer()
-  if (err != 0) {
-    s_log << "client_connect: Unable to set receive buffer size: "
-          << strerror(errno);
-    logger->Log(cantina::LogLevel::Critical, s_log.str());
-    throw std::runtime_error(s_log.str());
-  }
-
-  err =
-    setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &rcv_timeout, sizeof(rcv_timeout));
-  if (err != 0) {
-    s_log << "client_connect: Unable to set receive timeout: "
-          << strerror(errno);
-    logger->Log(cantina::LogLevel::Critical, s_log.str());
-    throw std::runtime_error(s_log.str());
-  }
-
-
-  struct sockaddr_in srv_addr;
-  memset((char*)&srv_addr, 0, sizeof(srv_addr));
-  srv_addr.sin_port = htons(serverInfo.port);
-  srv_addr.sin_family = AF_INET;
-  srv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
-
-  err = bind(fd, (struct sockaddr*)&srv_addr, sizeof(srv_addr));
-  if (err < 0) {
-    s_log << "connect_server: unable to bind to socket: " << strerror(errno);
-    logger->Log(cantina::LogLevel::Critical, s_log.str());
-    throw std::runtime_error(s_log.str());
-  }
-
-  logger->info << "connect_server: port: " << serverInfo.port << " fd: " << fd
-               << std::flush;
-
-  running_threads.emplace_back(&UDPTransport::fd_reader, this);
-  running_threads.emplace_back(&UDPTransport::fd_writer, this);
-
-  return last_context_id;
+TransportConnId UDPTransport::connect_server() {
+    std::stringstream s_log;
+
+    fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (fd < 0) {
+        s_log << "connect_server: Unable to create socket: " << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
+
+    // set for re-use
+    int one = 1;
+    int err =
+            setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (const char *) &one, sizeof(one));
+    if (err != 0) {
+        s_log << "connect_server: setsockopt error: " << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
+
+    // TODO: Add config for this value
+    size_t snd_rcv_max = 2000000;
+    timeval rcv_timeout{.tv_sec = 0, .tv_usec = 10000};
+
+    err =
+            setsockopt(fd, SOL_SOCKET, SO_SNDBUF, &snd_rcv_max, sizeof(snd_rcv_max));
+    if (err != 0) {
+        s_log << "client_connect: Unable to set send buffer size: "
+              << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
+
+    err =
+            setsockopt(fd, SOL_SOCKET, SO_RCVBUF, &snd_rcv_max, sizeof(snd_rcv_max));
+    if (err != 0) {
+        s_log << "client_connect: Unable to set receive buffer size: "
+              << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
+
+    err =
+            setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &rcv_timeout, sizeof(rcv_timeout));
```

Comment:
Is this needed on a UDP socket?

Reply:
It's used today so that we don't block on reading/waiting for data.

Reply:
This should put the socket in non-blocking mode:
```c
auto flags = fcntl(new_socket, F_GETFL);
fcntl(new_socket, F_SETFL, flags | O_NONBLOCK);
```

Each of those calls can return an error that needs to be checked.

Reply:
Right now we specifically do not want it in blocking mode. That's the reason we have the timeout and the dedicated threads for reader/writer. 

We can change it to non-blocking later now that we have a tick service that runs with very low impact on CPU and interrupts.